lib/waiter.js

```diff
@@ -0,0 +1,60 @@
+class Waiter {
+  constructor (promise, cb, expectReject) {
+    this.cb = cb
+    this.ready = false
+    this.value = null
+    this.resolved = false
+    this.rejected = false
+    this.done = false
+    this.finishing = false
+    this.expectReject = !!expectReject
+    this.promise = new Promise(res => this.resolve = res)
+    promise.then(value => {
+      if (this.done) {
+        return
+      }
+
+      this.resolved = true
+      this.value = value
+      this.done = true
+      this.finish()
+    }).catch(er => this.reject(er))
+  }
+
+  reject (er) {
+    if (this.done) {
+      return
+    }
```

Comment:
Similar to the resolve callback, this could overwrite the error reported to `.abort()`.